### PR TITLE
[breaking] Rewrite using atomic-fs-blob-store

### PR DESCRIPTION
**Pull request by gmaclennan**
This uses [atomic-fs-blob-store](https://github.com/binded/atomic-fs-blob-store) which in turn uses [fs-write-stream-atomic](https://github.com/npm/fs-write-stream-atomic) which has way more tests for broken writes, especially edge cases on windows.

I wrote more tests, which fail on the `flush` branch. This rewrite simplifies things a lot I think and should be more robust. I needed to use a fork of fs-write-stream-atomic to allow for custom names for tmp files, so that the `list()` function will work.

I also removed the `_` prefix from the `list()` function, since this is now a tested part of the API that we depend upon and should not be considered an internal function.

There are two failing tests because of a [breaking change](https://github.com/maxogden/abstract-blob-store/issues/15) that was introduced in the latest abstract-blob-store without a major version bump (not found file errors do not have a `notFound=true` property set)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,12 +15,14 @@
   },
   "keywords": [],
   "dependencies": {
-    "fs-blob-store": "^5.2.1",
+    "@digidem/atomic-fs-blob-store": "^5.3.0",
```
Comment (okdistribute) [inline package.json:18]: do we need to use a fork of this or can we use the npm published version that isn't digidem?

File: index.js [hunk 2]
```diff
@@ -1,96 +1,98 @@
...
-  return store.remove(opts, done)
+BlobStore.prototype.remove = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
+  opts.key = this._insertSubDirPrefix(opts.key)
+  return AtomicStore.prototype.remove.call(this, opts, cb)
 }
 
-// TODO: opts to choose whether to use staging area
-BlobStore.prototype.createWriteStream = function (opts, done) {
+BlobStore.prototype.list = function (cb) {
```
Comment (okdistribute) [inline index.js:71]: ._list with an underscore is depended upon by blob-store-replication stream so we'd need to update that code too https://github.com/noffle/blob-store-replication-stream/blob/master/index.js#L132

**Review by okdistribute — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (gmaclennan) [reply package.json:18]: We are Using that fork in order to use a fork of fs-create-write-stream-atomic which adds a new option to allow a custom naming function for the temp files. The opts are passed through by atomic-FS-blob-Store so if I get a PR accepted as a minor release it would just work. 
We need the custom naming function so that the list function works (so it doesn’t list temp files left by failed rights).

File: index.js (hunk 2 above)
Comment (gmaclennan) [reply index.js:71]: Yeah happy to leave this as a non-breaking change and keep as _write.

File: package.json (hunk 1 above)
Comment (okdistribute) [reply package.json:18]: sioundssounds good

File: index.js (hunk 2 above)
Comment (okdistribute) [reply index.js:71]: great, looks good to me then otherwise

**Comment by okdistribute** [comment]
@noffle if it looks okay to you can we merge and then change list to _list before publishing? this is on gregor's branch but I don't know if he'll have time to look at this since he's in the field

File: README.md [hunk 3]
```diff
@@ -36,6 +39,14 @@ ws.end(function() {
 })
 ```
 
+## API
+
+See https://github.com/maxogden/abstract-blob-store and in addition:
+
+### store.list(cb)
```
Comment (hackergrrl) [inline README.md:46]: You'll need to patch `blob-store-replication-stream` if you want this to work.

I suggest the patch checks for both `list` and `_list` and calls whatever it finds (`list` preferred), for backwards compatibility.

File: package.json (hunk 1 above)
Comment (hackergrrl) [inline package.json:18]: Bonus points if you send a PR to the original repo!

File: index.js [hunk 4]
```diff
@@ -1,96 +1,98 @@
...
+  return AtomicStore.prototype.createReadStream.call(this, key)
 }
 
-BlobStore.prototype.exists = function (opts, done) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.exists(opts, done)
+BlobStore.prototype.exists = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
```
Comment (hackergrrl) [inline index.js:60]: `if (typeof cb !== 'function') cb = noop` (or fail if not given)

File: index.js [hunk 5]
```diff
@@ -1,96 +1,98 @@
...
+  return AtomicStore.prototype.exists.call(this, opts, cb)
 }
 
-BlobStore.prototype.remove = function (opts, done) {
-  var name = typeof opts === 'string' ? opts : opts.key
-  var subdir = filenamePrefix(name, 7)
-  var store = this._getStore(subdir)
-  return store.remove(opts, done)
+BlobStore.prototype.remove = function (opts, cb) {
+  if (typeof opts === 'string') opts = {key: opts}
```
Comment (hackergrrl) [inline index.js:66]: `if (typeof cb !== 'function') cb = noop`

File: index.js (hunk 2 above)
Comment (hackergrrl) [reply index.js:71]: Leave it as `list` but also add `BlobStore.prototype._list = BlobStore.prototype.list`, and we can patch the replication stream module at our leisure.

File: index.js [hunk 6]
```diff
@@ -1,96 +1,98 @@
...
-      if (err) return done(err)
-      fs.rename(from, to, function (err) {
-        done(err, { key: name })
-      })
-    })
-  }
+BlobStore.prototype._insertSubDirPrefix = function (key) {
+  var prefixLen = this.subDirPrefixLen
+  var parsed = path.parse(key)
+  if (parsed.name.length <= prefixLen) return key
```
Comment (hackergrrl) [inline index.js:91]: The old code used as much of the string as possible as a prefix. I think that way will give better results than this: not doing any prefixing at all on short keys.

Comment (hackergrrl) [reply index.js:91]: I think you can just remove this one line and the `slice` below will work fine on short keys.

Comment (hackergrrl) [reply index.js:91]: This will also simplify the removal code below, since you know you can always just pop the top dir off.

**Review by hackergrrl — CHANGES_REQUESTED** [review]

File: index.js [hunk 7]
```diff
@@ -1,96 +1,98 @@
...
-function filenamePrefix (name, prefixLen) {
-  var extLen = path.extname(name).length
-  return name.substring(0, Math.min(prefixLen, name.length - extLen))
+BlobStore.prototype._removeSubDirPrefix = function (key) {
+  var prefixLen = this.subDirPrefixLen
+  var parsed = path.parse(key)
+  if (parsed.name.length <= prefixLen) return key
+  var dirs = parsed.dir.split('/')
+  if (parsed.name.slice(0, prefixLen) !== dirs.pop()) return key
+  return path.join(dirs.join('/'), parsed.base)
```
Comment (okdistribute) [inline index.js]: `path.sep` here for windows

File: index.js (hunk 5 above)
Comment (gmaclennan) [reply index.js:66]: This is passed through to atomic-fs-blob-store and handled there in accordance to abstract-blob-store specs

File: index.js (hunk 4 above)
Comment (gmaclennan) [reply index.js:60]: This is handled by atomic-fs-blob-store

File: index.js (hunk 6 above)
Comment (gmaclennan) [reply index.js:91]: I think there are edge-cases that could creep in (dot-files etc). I changed the way this is done now. There are maybe other ways of doing this but the current way I think I've thought through a few different edge cases.

Comment (hackergrrl) [reply index.js:91]: What does the subdir suffix add to this?

Can you say more about these edge cases you're thinking of? How do dotfiles cause problems?

Comment (gmaclennan) [reply index.js:91]: I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)

Comment (gmaclennan) [reply index.js:91]: I wrote a breaking test before this commit. The issue is a naming collision between two file named `fo` and `foobar`. Internally these collide (filename with same name as folder). 
The other issue if you are just using the prefix is two files with keys `foobar.txt` and `fo/obar.txt` which would overwrite each other. (This is with a súbdito prefix length of 2)

**Review by okdistribute — APPROVED** [review]

**Comment by hackergrrl** [comment]
Closing (merging #4 instead, which has some additional patches).
